### ISSUE TITLE
API: check balance + transfer to external address

### ISSUE THREAD

**Issue by kilrau** [raiden-network/raiden#4298]
## Abstract

For our use case, two API calls are missing, see specification.

## Motivation

We want to be able to 1) check our on-chain balance and 2) send on-chain funds to an external address without dealing with another client.

## Specification

1) check our own on-chain balance of ETH and token(s).
2) send on-chain ETH and Token(s) to an external address

Both should be relatively easy to expose since functionality already exists.

## Backwards Compatibility

It's new calls, I don't see a problem here.

**Comment by pirapira** [comment]
This sounds related to the "withdraw" feature that's currently being implemented https://github.com/raiden-network/raiden/pull/4130

**Comment by pirapira** [comment]
> send on-chain ETH and Token(s) to an external address

should this "send"ing happen using Raiden, or a usual gas-eating transfer?

**Comment by kilrau** [comment]
on-chain, gas-eating

**Comment by pirapira** [comment]
Then that sounds exactly like the "withdraw".

**Comment by LefterisJP** [comment]
> check our own on-chain balance of ETH and token(s).

Why does this need to happen through the Raiden API?
This is simple web3 calls to the ethereum chain.

> send on-chain ETH and Token(s) to an external address

The only reason for this is since we need to do nonce counting inside Raiden, right?

**Comment by pirapira** [comment]
Ah, maybe I was mistaken.  @kilrau is this issue about moving fund from a Raiden channel to somewhere?

**Comment by kilrau** [comment]
> Why does this need to happen through the Raiden API?

Good point. Doesn't *need* to since the functionality is there via a L1 web3 call, like you said. It just would be **nice**. It's just more convenient for projects building on top of Raiden to not have to deal with API's of (differing) lower layer clients, like an ethereum full node. Checking wallet/account balance is IMHO an essential feature that probably most projects building on top of Raiden need at some point, so it would be nice to have it exposed directly through the Raiden API. Especially since Raiden has the functionality already:
```
Your account's balance is below the estimated gas reserve of 0.003573350023965722 eth. This may lead to a loss of of funds because your account will be unable to perform on-chain transactions. Please add funds to your account as soon as possible.
```

> > send on-chain ETH and Token(s) to an external address
> 
> The only reason for this is since we need to do nonce counting inside Raiden, right?
> Ah, maybe I was mistaken. @kilrau is this issue about moving fund from a Raiden channel to somewhere?

Actually the same reason as above - abstracting from the lower layer client. And no, it's about sending tokens on-chain using gas to another external address on-chain. No channels involved at all. So if the withdrawal feature allows that (skimmed through the PR, but was a bit hard to find out if it only supports channel-related withdrawals), we are good here.

In summary do the two features above provide basic *wallet* functionality

But this is just my product view on things and strongly biased by how we handle API's with `xud` by exposing essential functionality through `xud`'s gRPC to abstract away from L2 and L1 clients ✌️

**Comment by rakanalh** [comment]
> The only reason for this is since we need to do nonce counting inside Raiden, right?

Yes. Otherwise, Raiden will crash / behave inconsistently if the private key was used outside to perform a transaction on-chain to transfer funds or anything else for that matter.

I think that as long as the "private key should only be used by Raiden" rule stands, basic on-chain functions such as transferring ETH from one account to another should be provided by Raiden.

However, since this feature is "nice to have", then the priority of this needs to be determined.

**Comment by kilrau** [comment]
Also we noticed different kind of errors when we moved funds out of the raiden keystore file. Restarting raiden solved it. We should investigate the reason for this, and will details of the errors in the course of this issue.

**Comment by LefterisJP** [comment]
@kilrau If that happened while raiden was running this is a known issue. It has to do with the nonce handling.

**Comment by kilrau** [comment]
Can we bump up prio on this one?

**Comment by LefterisJP** [comment]
> Can we bump up prio on this one?

Hey Kilian, this definitely won't be in Alderaan, but perhaps we can put it in a minor release after it.

Is it that important to you?

**Comment by kilrau** [comment]
> we can put it in a minor release after it.

That works.

> Is it that important to you?

Yes, directly after https://github.com/raiden-network/raiden/issues/3945 /  https://github.com/raiden-network/raiden-contracts/issues/1216 (which enables multi-hop channels for our use case).

Why is it important: our users are managing all ETH funds with *one* keystore file generated by `xud` and governed by raiden. Without above functionality there is just no clean way to move ETH funds on-chain for our users.

**Comment by LefterisJP** [comment]
All right.

But one question that also @rakanalh brought up.

> unlock a keystore file (not via password file, this is security-wise problematic: ExchangeUnion/xud#1247 (review)), something like v1/api/unlock - { address: "...", password: "pswd" }

Why do you need the unlocking of the account as a separate action? That will have no benefit for you.

**Comment by kilrau** [comment]
Fair question. Let's look at the facts: the raiden client can't use the eth keystore/account after start, it has to be "unlocked" first. The question is *how* we want to unlock.

Reasons why we think a separate `unlock` api call similar to https://api.lightning.community/#unlockwallet makes sense:
* Storing a password file with the password to unlock the keystore file in the same directories is not only inconvenient in case of changes but also somewhat defeats the purpose of the password.
* In order to make life easier for our users, we have a so-called master password feature, where `xud` creates one seed for all L2 clients and secures all L2 wallets with the same password. On start, `xud` asks user to enter this password and programatically unlocks L2 wallets via API call. This is currently not possible with raiden and we would need to store this password permanently in the password file and raiden would always unlock automatically. The result would be user entering a wrong password in xud still having full access to the eth keystore/account.

**Comment by LefterisJP** [comment]
Raiden has no concept of unlocking a key during the run. Once it starts it unlocks and keeps the unlocked private key in memory.

And from then and on raiden signs all transactions, and messages itself and then broadcasts it to either the ethereum network or sends its to its off-chain partners via matrix.
